### PR TITLE
Add support for arbitrary Python iterables in preprocessing layer adapt() methods

### LINKED ISSUE #22156: Support arbitrary Python iterables in preprocessing layer adapt() methods
Currently, most preprocessing layers (`FeatureSpace`, `IndexLookup`, `TextVectorization`, `Discretization`) only accept `tf.data.Dataset` objects in their `adapt()` methods. This creates friction for users working with JAX or PyTorch backends who typically have data in other formats like lists, generators, or custom iterables.

### Current Limitation

```python
# This works
import tensorflow as tf
dataset = tf.data.Dataset.from_tensor_slices(data)
layer.adapt(dataset)

# This doesn't work
layer.adapt(["apple", "banana", "cherry"])  # Error
```

## Proposed Solution

Enable preprocessing layers to accept arbitrary Python iterables (lists, generators, custom iterables) in their `adapt()` methods, similar to how `Normalization` already works.

### Expected Behavior

```python
# Lists should work
layer.adapt(["apple", "banana", "cherry"])  

# Generators should work (memory efficient!)
def data_generator():
    for item in huge_dataset:
        yield item
layer.adapt(data_generator())  

# Custom iterables should work
layer.adapt(custom_iterable_object)  
```

## Benefits

1. **Backend Agnostic**: JAX and PyTorch users can use native Python data structures without converting to TensorFlow datasets
2. **Memory Efficient**: Generators allow processing large datasets without loading everything into memory
3. **Consistent API**: All preprocessing layers would behave consistently with `Normalization`
4. **Backward Compatible**: Existing code using `tf.data.Dataset` continues to work
5. **Pythonic**: More natural for users coming from standard Python/NumPy workflows

## Affected Layers

- `FeatureSpace`
- `StringLookup` / `IntegerLookup` (via `IndexLookup`)
- `TextVectorization`
- `Discretization`

### 1. Processing large text files with generators
```python
def text_generator():
    with open("large_corpus.txt") as f:
        for line in f:
            yield line.strip()

vectorizer = TextVectorization()
vectorizer.adapt(text_generator(), batch_size=64)
```

### 2. JAX/PyTorch users with native data structures
```python
# JAX users with lists
data = ["text1", "text2", "text3"]
layer.adapt(data)

# PyTorch users with custom iterables
layer.adapt(pytorch_dataloader)
```

### 3. Quick prototyping and testing
```python
# Simple list for quick testing
lookup = StringLookup()
lookup.adapt(["cat", "dog", "bird"])
```

## Related

This aligns with Keras 3's goal of being backend-agnostic and improving multi-backend compatibility. The idea for this came as talked about in /keras-team/keras/issues/18442.

### PR DESCRIPTION
**Pull request by goyaladitya05**
This PR enables preprocessing layers to accept arbitrary Python iterables (lists, generators, custom iterables) in their [adapt()](cci:1://file:///d:/codes/open-source/keras/keras/src/layers/preprocessing/discretization.py:158:4-196:29) methods, making them more flexible and backend-agnostic.

Currently, most preprocessing layers only accept `tf.data.Dataset` objects in their [adapt()](cci:1://file:///d:/codes/open-source/keras/keras/src/layers/preprocessing/discretization.py:158:4-196:29) methods. This limits usability for users working with JAX or PyTorch backends who often have data in other formats. This change addresses the feature request in #18442.

### Modified Files

- **[feature_space.py](cci:7://file:///d:/codes/open-source/keras/keras/src/layers/preprocessing/feature_space.py:0:0-0:0)**: Enhanced `FeatureSpace.adapt()` to accept generators and custom iterables yielding dicts
- **[index_lookup.py](cci:7://file:///d:/codes/open-source/keras/keras/src/layers/preprocessing/index_lookup.py:0:0-0:0)**: Updated `IndexLookup.adapt()` to handle iterables (benefits `StringLookup` and `IntegerLookup`)
- **[text_vectorization.py](cci:7://file:///d:/codes/open-source/keras/keras/src/layers/preprocessing/text_vectorization.py:0:0-0:0)**: Improved `TextVectorization.adapt()` with batch processing for iterables
- **[discretization.py](cci:7://file:///d:/codes/open-source/keras/keras/src/layers/preprocessing/discretization.py:0:0-0:0)**: Extended `Discretization.adapt()` to handle generators and iterables

### New Files

- **[iterable_adapt_test.py](cci:7://file:///d:/codes/open-source/keras/keras/src/layers/preprocessing/iterable_adapt_test.py:0:0-0:0)**: Test suite with 14 tests covering all scenarios

### Before
```python
import tensorflow as tf
dataset = tf.data.Dataset.from_tensor_slices(data)
layer.adapt(dataset)
```

### After
```python
# Lists
layer.adapt(["apple", "banana", "cherry"])

# Generators (memory efficient!)
def data_generator():
    for item in huge_dataset:
        yield item
layer.adapt(data_generator())

# Custom batch size for TextVectorization
def text_gen():
    with open("large_file.txt") as f:
        for line in f:
            yield line
layer.adapt(text_gen(), batch_size=64)
```

Closes #22156

### CODE REVIEW

**Comment by goyaladitya05** [comment]
All CI checks are now passing across all backends.

@gbaned @hertschuh , this PR is ready for review whenever you have the time. Thanks!

File: keras/src/layers/preprocessing/iterable_adapt_test.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+from keras import utils
+from keras.src import layers
+from keras.src import testing
+
+
+class IterableAdaptTest(testing.TestCase, parameterized.TestCase):
+    @parameterized.parameters([("list",), ("generator",)])
+    def test_feature_space_adapt_with_iterables(self, input_type):
+        def create_data():
```
Comment (hertschuh) [inline keras/src/layers/preprocessing/iterable_adapt_test.py:14]: These tests should be moved to each respective existing test file.

**Comment by goyaladitya05** [comment]
Closing it as https://github.com/keras-team/keras/pull/22417 was merged for the feature.
